python/ray/data/block.py

```diff
@@ -158,30 +162,89 @@ def __repr__(self):
     called, the time delta is saved as part of the stats.
     """
 
-    def __init__(self):
-        self.start_time = time.perf_counter()
-        self.start_cpu = time.process_time()
+    def __init__(self, poll_interval_s: Optional[float] = None):
+        """
+
+        Args:
+            poll_interval_s: The interval to poll the RSS of the process. If `None`,
+                this class won't poll the RSS.
+        """
+        self._poll_interval_s = poll_interval_s
+
+        # Record start times.
+        self._start_time = time.perf_counter()
+        self._start_cpu = time.process_time()
+
+        # Record initial RSS.
+        self._process = psutil.Process(os.getpid())
+        self._max_rss = int(self._process.memory_info().rss)
+
+        # If necessary, start the RSS poll thread.
+        self._rss_poll_thread = None
+        self._stop_rss_poll_event = None
+        if self._poll_interval_s is not None:
+            (
+                self._rss_poll_thread,
+                self._stop_rss_poll_event,
+            ) = self._start_rss_poll_thread()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if self._rss_poll_thread is not None:
+            self._stop_rss_poll_thread()
 
     def build(self) -> "BlockExecStats":
-        self.end_time = time.perf_counter()
-        self.end_cpu = time.process_time()
+        # Record end times.
+        end_time = time.perf_counter()
+        end_cpu = time.process_time()
+
+        # Record max RSS.
+        self._max_rss = max(self._max_rss, int(self._process.memory_info().rss))
 
+        # Build the stats.
         stats = BlockExecStats()
-        stats.start_time_s = self.start_time
-        stats.end_time_s = self.end_time
-        stats.wall_time_s = self.end_time - self.start_time
-        stats.cpu_time_s = self.end_cpu - self.start_cpu
-        process = psutil.Process(os.getpid())
-        stats.rss_bytes = int(process.memory_info().rss)
+        stats.start_time_s = self._start_time
+        stats.end_time_s = end_time
+        stats.wall_time_s = end_time - self._start_time
+        stats.cpu_time_s = end_cpu - self._start_cpu
+        stats.max_rss_bytes = self._max_rss
 
         return stats
 
+    def reset(self):
+        self._start_time = time.perf_counter()
+        self._start_cpu = time.process_time()
+        self._max_rss = int(self._process.memory_info().rss)
+
+    def _start_rss_poll_thread(self) -> Tuple[threading.Thread, threading.Event]:
+        assert self._poll_interval_s is not None
+
+        stop_event = threading.Event()
+
+        def poll_rss():
+            while not stop_event.is_set():
+                rss_bytes = int(self._process.memory_info().rss)
+                self._max_rss = max(self._max_rss, rss_bytes)
+                stop_event.wait(self._poll_interval_s)
+
+        thread = threading.Thread(target=poll_rss)
```

Comment:
nit, maybe add deamon=True, so the thread won't prevent the process to stop (for whatever reason).